packages/http-server/src/__tests__/server.oas.spec.ts

```diff
@@ -228,6 +253,67 @@ describe.each([['petstore.oas2.json'], ['petstore.oas3.json']])('server %s', fil
     }
   });
 
+  describe('server validation: given __server query param', () => {
+    test('when the server is not valid then return error', async () => {
+      const response = await server.fastify.inject({
+        method: 'GET',
+        url: '/pets/10?__server=https://google.com',
+      });
+
+      expect(response.statusCode).toBe(404);
+      expect(response.payload).toEqual(
+        '{"type":"https://stoplight.io/prism/errors#NO_SERVER_MATCHED_ERROR","title":"Route not resolved, no server matched.","status":404,"detail":"The base url https://google.com hasn\'t been matched with any of the provided servers"}',
+      );
+    });
+
+    test('when the server is valid then return 200', async () => {
+      const response = await server.fastify.inject({
+        method: 'GET',
+        url: '/pets/10?__server=https://petstore.swagger.io/v2',
+      });
+
+      expect(response.statusCode).toBe(200);
+    });
+
+    // oas2 does not support overriding servers
+    if (file === 'petstore.oas3.json') {
+      describe('and operation overrides global servers', () => {
+        test(`when the server is valid then return 200`, async () => {
+          const response = await server.fastify.inject({
+            method: 'GET',
+            url: '/store/inventory?__server=https://petstore.swagger.io/v3',
+          });
+
+          expect(response.statusCode).toBe(200);
+        });
+
+        test(`when the server is not valid for this exact operation then return error`, async () => {
+          const response = await server.fastify.inject({
+            method: 'GET',
+            url: '/store/inventory?__server=https://petstore.swagger.io/v2',
+          });
+
+          expect(response.statusCode).toBe(404);
+          expect(response.payload).toEqual(
+            '{"type":"https://stoplight.io/prism/errors#NO_SERVER_MATCHED_ERROR","title":"Route not resolved, no server matched.","status":404,"detail":"The base url https://petstore.swagger.io/v2 hasn\'t been matched with any of the provided servers"}',
+          );
+        });
+
+        test(`when the server is invalid return error`, async () => {
+          const response = await server.fastify.inject({
+            method: 'GET',
+            url: '/store/inventory?__server=https://notvalid.com',
+          });
```

Comment:
This all looks great! An idea for an extra rest: can we define a server with variables then see if a URL with a valid and invalid server according to those validations performs as expected?

Reply:
Sure! Good idea, will add it in a moment.
It's actually interesting to know that we have test cases covering this :) 
FYI each package has its own set of unit + integration tests. E.g. testing that a server matches is currently tested with: a unit test testing the actual matcher function implementation, a unit test testing some higher level component using that function, an integration test testing the Http Client package and now an integration test in the Http Server package :D 

I have to add it to the test-harness (e2e) too :)

Reply:
Btw I've been thinking about this for a while now and a lot of the integration test "kind of" repeat. 
For instance I can have 10 integration tests checking server validation at a Http Client level (`http-prism-instance.spec.ts`).
I could have 10 exact same integrate tests at a Http Serevr level (`server.oas.spec.ts`).
Finally one could argue that we should have the same 10 tests a the test harness level...

But, even though, they sometimes spot the issue I think it's an overkill. Any thoughts on how to stay sane?

Reply:
Yep there is crossover of some things, but it usually goes a little like this.

![image](https://user-images.githubusercontent.com/67381/59435232-7028d300-8ded-11e9-825f-09b9e29d88de.png)

You will not want to test every single possible thing in every single possible way in the test harness, but we will be testing some of the same things sure. 

And the HTTP Server and HTTP Client both integrate with the mocker logic. We need to make sure that integration works. Yes.